### ISSUE TITLE
Add mTLS client attributes to matching/routing 

### ISSUE THREAD

**Issue by dekimsey** [mholt/caddy-l4#66]
The http app exposes a number of placeholders that'd be equally valuable for the L4 side (`http.request.tls.*`). Right now matching only works for `sni`, `alpn`, and `remote_ip`.

Use-case:
I'd like to be able to have an mTLS stripping proxy that can also validate the client's certificate has a given SAN in addition to being validly signed by the CA. Right now, that doesn't appear possible.

(Apologies if I have screwed up the nomenclature here, I'm a bit new to Caddy)

**Comment by mholt** [comment]
Yeah, we can do that. This is pretty easy to do. You're looking for basically this code:

https://github.com/caddyserver/caddy/blob/2642bd72b7ca35b8622824fdffced2aefe1aaf11/modules/caddyhttp/replacer.go#L286-L379

But for layer4. The Connections all have a Replacer, similar to the http app:

https://github.com/mholt/caddy-l4/blob/aec6535658b117e6283cc2bdaa80d17462fd2eef/layer4/connection.go#L32-L46

So if you're matching on TLS attributes (as opposed to terminating TLS) then maybe somewhere around here:

https://github.com/mholt/caddy-l4/blob/aec6535658b117e6283cc2bdaa80d17462fd2eef/modules/l4tls/matcher.go#L101-L103

I'm a little too busy to take this on right now but it shouldn't be too difficult if someone would like to contribute!

**Comment by dekimsey** [comment]
Thanks for the pointers @mholt, I'll give this a try.

**Comment by vnxme** [comment]
The first part of this issue is completed with #67. See also the [docs](https://github.com/mholt/caddy-l4/blob/master/docs/handlers/tls.md) for details on the placeholders registered by the `tls` handler.

The other part is enabling the Layer 4 app to do matching and routing based on these placeholders. It's already [possible](https://github.com/mholt/caddy-l4/blob/master/docs/examples/tls_sni_dynamic_upstreams.md) to use placeholders in the dial address field of proxy upstreams, but it's not enough. I have an idea about how to implement placeholder-driven routing -- I will submit a separate PR dealing with that.

**Comment by vnxme** [comment]
The other part has been added in #396. Now it's possible to route traffic based on any placeholders and custom variables, including TLS client certificate fields.

E.g. use `vars {l4.tls.client.san.dns_names.0} test.example.com` to match against the first  SAN of the domain name type in the client's certificate. 

Closing this issue as completed. Feel free to reopen it or start a new discussion.
